### PR TITLE
Expose option to load vector tile sources from files in Data Source Manager

### PR DESCRIPTION
**Pull request by nyalldawson**
The Vector Tile tab now includes choices for selecting from a "Service" (the existing, connection based approach for adding
vector tiles from online sources) or "File" (the new option which allows directly adding VTPK or MBTiles vector tile files)

https://user-images.githubusercontent.com/1829991/229019730-664efa91-41a4-4dc6-ba34-6ab573979047.mp4

Funded by Landesamt für Vermessung und Geoinformation, Feldkirch, Austria

### CODE REVIEW

File: src/app/qgshandlebadlayers.cpp [hunk 1]
```diff
@@ -270,28 +270,29 @@ void QgsHandleBadLayers::browseClicked()
     const QString provider = mLayerList->item( row, 0 )->data( static_cast< int >( CustomRoles::Provider ) ).toString();
 
     QString fileFilter;
+    QgsProviderMetadata *metadata = QgsProviderRegistry::instance()->providerMetadata( provider );
```
Comment (nirvn) [inline src/app/qgshandlebadlayers.cpp:273]: There's zero chance here we'd end up with a null pointer (e.g. provider not having a registered metadata counterpart)?

File: src/app/qgshandlebadlayers.cpp [hunk 2]
```diff
@@ -270,28 +270,29 @@ void QgsHandleBadLayers::browseClicked()
...
-        fileFilter = QgsProviderRegistry::instance()->providerMetadata( provider )->filters( QgsProviderMetadata::FilterType::FilterRaster );
+        fileFilter = metadata->filters( Qgis::FileFilterType::Raster );
         break;
       case Qgis::LayerType::Mesh:
         memoryQualifier = QStringLiteral( "lastMeshFileFilter" );
         fileFilter = QgsProviderRegistry::instance()->fileMeshFilters();
         break;
       case Qgis::LayerType::VectorTile:
         memoryQualifier = QStringLiteral( "lastVectorTileFileFilter" );
-        // not quite right -- but currently there's no generic method to get vector tile filters...
```
Comment (nirvn) [inline src/app/qgshandlebadlayers.cpp:289]: :heart: nice to see this comment made obsolete :)

File: src/core/mesh/qgsmesheditor.cpp [hunk 3]
```diff
@@ -22,6 +22,7 @@
 #include "qgsgeometryengine.h"
 #include "qgsmeshadvancedediting.h"
 #include "qgsgeometryutils.h"
+#include "qgspolygon.h"
```
Comment (nirvn) [inline src/core/mesh/qgsmesheditor.cpp:25]: Needed? Just checking.

File: src/core/providers/qgsproviderregistry.cpp [hunk 4]
```diff
@@ -411,6 +414,24 @@ void QgsProviderRegistry::init()
       }
     }
 
+    // now get vector tile file filters, if any
+    const QString fileVectorTileFilters = meta->filters( Qgis::FileFilterType::VectorTile );
+    if ( !fileVectorTileFilters.isEmpty() )
+    {
+      QgsDebugMsgLevel( "vector tile filters: " + fileVectorTileFilters, 2 );
```
Comment (nirvn) [inline src/core/providers/qgsproviderregistry.cpp:421]: Maybe worth adjusting debug message so it matches preexisting ones?

File: src/gui/vectortile/qgsvectortilesourceselect.cpp [hunk 5]
```diff
@@ -65,6 +80,15 @@ QgsVectorTileSourceSelect::QgsVectorTileSourceSelect( QWidget *parent, Qt::Windo
...
 
   populateConnectionList();
+
+  mFileWidget->setDialogTitle( tr( "Open Vector Tile Dataset" ) );
+  mFileWidget->setFilter( QgsProviderRegistry::instance()->fileVectorTileFilters() );
+  mFileWidget->setStorageMode( QgsFileWidget::GetFile );
+  mFileWidget->setOptions( QFileDialog::HideNameFilterDetails );
+  connect( mFileWidget, &QgsFileWidget::fileChanged, this, [ = ]( const QString & path )
+  {
+    emit enableButtons( ! path.isEmpty() );
```
Comment (nirvn) [inline src/gui/vectortile/qgsvectortilesourceselect.cpp:90]:
```suggestion
    emit enableButtons( !path.isEmpty() );
```

**Review by nirvn — APPROVED** [review]
Just a few comments / words of appreciation :wink:

File: src/app/qgshandlebadlayers.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/app/qgshandlebadlayers.cpp:273]: I'm not sure -- you'll notice I left the existing logic for all non vector tile layers, but added a nullptr check for the new vector tile handling. I figured there's been no bug reports for the other layer types, so the non-null assumption must be valid for those!

File: src/core/mesh/qgsmesheditor.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/core/mesh/qgsmesheditor.cpp:25]: Yep, the header cleanup elsewhere in this pr changes things here.

File: src/core/providers/qgsproviderregistry.cpp (hunk 4 above)
Comment (nyalldawson) [reply src/core/providers/qgsproviderregistry.cpp:421]: It's an exact match for the point cloud one just above! (I copy pasted all that code 😂)

Comment (nirvn) [reply src/core/providers/qgsproviderregistry.cpp:421]: @nyalldawson , doh! I was looking at the ones further up. No worries then, let it be :)

**Pull request merged**